src/msg/async/frames_v2.h

```diff
@@ -91,8 +114,21 @@ struct preamble_block_t {
 static_assert(sizeof(preamble_block_t) % CRYPTO_BLOCK_SIZE == 0);
 static_assert(std::is_standard_layout<preamble_block_t>::value);
 
+// V2 epilogue conveys integrity/authentication information. It's added
+// at the end of each frame holds:
+//  * CRC32 for MAX_NUM_SEGMENTS -- in plain mode,
+//  * cipher-specific data (e.g. auth tag for AES-GCM).
+union epilogue_block_t {
+  // TODO: add crc32 for plain mode
+  char auth_tag[CRYPTO_BLOCK_SIZE];
+  std::array<__le32, MAX_NUM_SEGMENTS> crc_values;
```

Comment:
we also need a flags field in the epilogue (or footer?) for the message aborts, and that is needed for both crc and secure mode.

Reply:
that is a TAG_MESSAGE specific thing (that's the only time we need to abort a frame send) so we could add a final "footer" segment that has the flags...

Reply:
I expect we want to have this information as close to frame's end as possible to avoid tx buffering. If it had been available at preamble crafting, we wouldn't have started transmitting at all, I guess.

Let me summarize the approaches:
* extra bit in `epilogue`. This would inflate the frame size but maybe just by one byte. The epilogue part will never be encrypted, so aligning to cipher's block size doesn't look necessary here IIUC.
* Extra segment (called e.g. *footer*) just for `Message`. I'm inclined to avoid doing that as we don't currently have room for fifth one (`Message` needs 4 segments at the moment due to `ceph_msg_header2`).

Reply:
Added a commit with proposal of the `flags` field being a part of `epilogue_block_t`.

Reply:
Hmm, it seems right for the crc mode, but for secure mode, is there a way to include it in the encrypted stream without padding things out extensively?  Appended at the end like this it's neither encrypted nor checked for integrity, so a MITM can selectively drop messages

Reply:
Just for GCM we could pick e.g. 120 bits as the length of auth tag but that way flexibility for potential cipher exchange in the future would be limited even more (moving to fixed-size *epilogue* had already constrained `sizeof(auth_tag)`, I'm afraid).

I'm not sure whether gluing the CRCs with auth tag and *late flags* is conceptually the right thing to do. If we somehow could ignore (or lower importance) of `ceph_msg_header::data_off`, following alternative becomes imaginable:
* make the *late flags* a part of `ceph_msg_header2`,
* transmit `ceph_msg_header2` **after** `front`, `middle`, and `data`.
* in crc mode use fixed-size epilogue, in secure allow cipher to determine it.

Reply:
why can't we just add the flags byte to the payload of the last segment that contains data?

Reply:
> why can't we just add the flags byte to the payload of the last segment that contains data?

That's also an option with its own consequences. One is blurring the distinction between data and metadata. Thus, the *last segment* would become **special**. I'm afraid about inflating allocation sizes -- if `data` is e.g. 4 KiB chunk, 4097 bytes would be allocated. Alternatively, we could go with two separated reads (4096 + 1) but this puts more complexity to the *last segment handler* and makes it even more special.

Reply:
The problem with moving the msg header to the last segment is that we then wouldn't get teh tid before the data payload, which precludes an rx_buffers-like zero-copy implementation.

It really seems like we need an additional small footer segment...

Reply:
Before going further, let me verify the design invariants/requirements: is avoiding 15 bytes of overhead worth the complexity inflation?

Reply:
I don't understand why we can't do something simple like:

* for *crc*: add the flags to an epilogue_crc_t (that includes flags and a vector of crcs).  the is an assumed part of the frame sequenc eon the wire: preamble, segment(s), epilogue_crc.
* for *secure*: make an epilogue_secure_t (that includes just flags), followed by the cipher's trailing epilogue_secure_cipher_t.

Reply:
Minutes from the stand-up:
* for now we can go with just moving the message header. `data_off` is used solely for enforcing alignment while `rx_buffers` are knocked out for a while.
* Bringing back 0c will need altering preamble format. We have 2 bytes reserved for flags/versioning.
* Determining secure signature size will be kept at the cipher's side.